src/memory_manager.rs

```diff
@@ -312,21 +312,98 @@ pub fn is_live_object(object: ObjectReference) -> bool {
+/// Return false otherwise.  This function does not panic.
+///
+/// This function is useful for conservative root scanning.  It can take any word-sized value,
+/// usually from the stack, and check whether it happens to look like an object reference.
+///
+/// Note: If the VM space (enabled by the `vm_space` feature) is present, it is currently
+/// unspecified whether the return value will be true or false if `addr` is within the VM space.
+/// The alloc bit is currently not availabe for the VM space.  In the future, we will require the
+/// VM to provide a bitmap to identify where objects are in the boot image.  After that, this
+/// function will be precise for the VM space, too.
+///
+/// If your VM uses both the VM space and conservative stack scanning, keep in mind that this
+/// function cannot identify objects in the VM space.  One way to work around this problem is to
+/// ensure objects in the VM space cannot move, and treat objects (or object reference fields) in
+/// the VM space as roots.  By doing this, there is no need to trace references to objects in the
+/// VM space.
+///
+/// Argument:
+/// * `addr`: An arbitrary address.
+#[cfg(feature = "is_mmtk_object")]
+pub fn is_mmtk_object<VM: VMBinding>(addr: Address) -> bool {
+    crate::util::is_mmtk_object::is_mmtk_object::<VM>(addr)
+}
+
+/// This function is the same as `is_mmtk_object`, except that the caller must ensure that
+/// -   `addr` is not 0, and
+/// -   `addr` is properly aligned for the object reference in the VM.
+///     -   i.e. `crate::util::is_mmtk_object::is_aligned_for_alloc_bit(addr)` is true. See the
+///         source code.
+/// Otherwise the behavior is undefined.
+///
+/// Most words on the stack are non-reference values.  For performance reasons, it is helpful if
+/// the VM filters out most words that are obviously not object references before calling into
+/// MMTk API.  *Null pointers* and *misaligned pointer* are obvious invalid.  The VM is also
+/// encouranged to filter out invalid object references using other heuristics, including:
+/// -   addresses out of the range allowed by the architecture or the operating system
+///     -   e.g. on Linux on x86_64, the virtual memory area for user-space programs is from
+///         `0x0000_0000_0000_0000` to `0x0000_7fff_ffff_ffff`.  Address outside this range are
+///         obviously not pointing to any MMTk objects.
+///     -   Note that GC Algorithms like ZGC use high bits of a pointer as tags, and may break the
+///         assumption above.
+/// -   addresses out of the MMTk heap range
+///     -   The MMTk heap range can be obtained from the `starting_heap_address()` and the
+///         `last_heap_address()` API functions.  But currently they do not work for MallocSpace.
+///
+/// Arguments:
+/// * `addr`: An address. Must not be 0, and must be properly aligned for the VM.
+#[cfg(feature = "is_mmtk_object")]
+pub fn is_mmtk_object_prechecked<VM: VMBinding>(addr: Address) -> bool {
```

Comment:
Yes. That's reasonable.

I introduced both because I think the semantics of `is_mmtk_object_prechecked` is a bit hard to express without the plain `is_mmtk_object`.  But from your comments below, I think we'd better leave it to the VM.  We will still require the VM to guarantee that there cannot be two objects competing the same alloc bit.